File: go/devices_test.go
package nitroping_test

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/url"
	"testing"

	nitroping "github.com/productdevbook/nitroping-sdk/go"
)

func TestDevicesRegister_PostsSnakeCaseBody(t *testing.T) {
	var captured struct {
		method string
		path   string
		auth   string
		body   map[string]any
	}

	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		captured.method = r.Method
		captured.path = r.URL.Path
		captured.auth = r.Header.Get("Authorization")
		raw, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(raw, &captured.body)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"id":"dev-1","created":true}`))
	})

	result, err := client.Devices.Register(context.Background(), nitroping.DeviceRequest{
		Platform: nitroping.PlatformIOS,
		Token:    "apns-token-abc",
		UserID:   nitroping.String("user-42"),
		Metadata: map[string]any{"source": "tests"},
	})
	if err != nil {
		t.Fatalf("Register: %v", err)
	}
	if result.ID != "dev-1" || !result.Created {
		t.Errorf("got %+v, want {dev-1 true}", result)
	}
	if captured.method != "POST" {
		t.Errorf("method = %q, want POST", captured.method)
	}
	if captured.path != "/api/v1/devices" {
		t.Errorf("path = %q, want /api/v1/devices", captured.path)
	}
	if captured.auth != "ApiKey np_test_secret" {
		t.Errorf("Authorization = %q", captured.auth)
	}
	if captured.body["platform"] != "ios" {
		t.Errorf("body.platform = %v", captured.body["platform"])
	}
	if captured.body["token"] != "apns-token-abc" {
		t.Errorf("body.token = %v", captured.body["token"])
	}
	if captured.body["user_id"] != "user-42" {
		t.Errorf("body.user_id = %v", captured.body["user_id"])
	}
	meta, _ := captured.body["metadata"].(map[string]any)
	if meta["source"] != "tests" {
		t.Errorf("body.metadata = %v", captured.body["metadata"])
	}
}

func TestDevicesRegister_WebPlatformIncludesKeys(t *testing.T) {
	var captured map[string]any
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(raw, &captured)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"id":"dev-2","created":false}`))
	})

	result, err := client.Devices.Register(context.Background(), nitroping.DeviceRequest{
		Platform:      nitroping.PlatformWeb,
		Token:         "https://fcm.googleapis.com/abc",
		WebPushP256dh: nitroping.String("BPS_p256dh_value"),
		WebPushAuth:   nitroping.String("auth_secret_value"),
	})
	if err != nil {
		t.Fatalf("Register: %v", err)
	}
	if result.Created {
		t.Errorf("Created = true, want false")
	}
	if captured["web_push_p256dh"] != "BPS_p256dh_value" {
		t.Errorf("web_push_p256dh = %v", captured["web_push_p256dh"])
	}
	if captured["web_push_auth"] != "auth_secret_value" {
		t.Errorf("web_push_auth = %v", captured["web_push_auth"])
	}
}

func TestDevicesRegister_TagsForwarded(t *testing.T) {
	var captured map[string]any
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(raw, &captured)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"id":"dev-3","created":true}`))
	})

	_, err := client.Devices.Register(context.Background(), nitroping.DeviceRequest{
		Platform: nitroping.PlatformAndroid,
		Token:    "fcm-token",
		Tags:     []string{"beta", "vip"},
	})
	if err != nil {
		t.Fatalf("Register: %v", err)
	}
	tags, _ := captured["tags"].([]any)
	if len(tags) != 2 || tags[0] != "beta" || tags[1] != "vip" {
		t.Errorf("body.tags = %v, want [beta vip]", captured["tags"])
	}
}

func TestDevicesUpdate_PutsTags(t *testing.T) {
	var captured struct {
		method string
		path   string
		auth   string
		ctype  string
		body   map[string]any
	}
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		captured.method = r.Method
		captured.path = r.URL.Path
		captured.auth = r.Header.Get("Authorization")
		captured.ctype = r.Header.Get("Content-Type")
		raw, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(raw, &captured.body)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"id":"dev-1","tags":["beta","vip"]}`))
	})

	result, err := client.Devices.Update(context.Background(), "dev-1", nitroping.UpdateDeviceRequest{
		Tags: []string{"beta", "vip"},
	})
	if err != nil {
		t.Fatalf("Update: %v", err)
	}
	if captured.method != "PUT" {
		t.Errorf("method = %q, want PUT", captured.method)
	}
	if captured.path != "/api/v1/devices/dev-1" {
		t.Errorf("path = %q", captured.path)
	}
	if captured.auth != "ApiKey np_test_secret" {
		t.Errorf("Authorization = %q", captured.auth)
	}
	if captured.ctype != "application/json" {
		t.Errorf("Content-Type = %q", captured.ctype)
	}
	tags, _ := captured.body["tags"].([]any)
	if len(tags) != 2 || tags[0] != "beta" || tags[1] != "vip" {
		t.Errorf("body.tags = %v, want [beta vip]", captured.body["tags"])
	}
	if result.ID != "dev-1" {
		t.Errorf("result.ID = %q, want dev-1", result.ID)
	}
	if len(result.Tags) != 2 || result.Tags[0] != "beta" || result.Tags[1] != "vip" {
		t.Errorf("result.Tags = %v, want [beta vip]", result.Tags)
	}
}

func TestDevicesUpdate_404APIError(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"error":{"code":"not_found","message":"Device not found"}}`))
	})

	_, err := client.Devices.Update(context.Background(), "missing", nitroping.UpdateDeviceRequest{
		Tags: []string{"x"},
	})
	if err == nil {
		t.Fatal("expected APIError")
	}
	var apiErr *nitroping.APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if apiErr.StatusCode != 404 {
		t.Errorf("StatusCode = %d", apiErr.StatusCode)
	}
	if apiErr.Code != "not_found" {
		t.Errorf("Code = %q", apiErr.Code)
	}
}

func TestDevicesUpdate_EmptyIDReturnsError(t *testing.T) {
	client, err := nitroping.NewClient(nitroping.ClientOptions{APIKey: "np_x"})
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	_, err = client.Devices.Update(context.Background(), "", nitroping.UpdateDeviceRequest{})
	if err == nil {
		t.Fatal("expected error on empty device id")
	}
}

func TestDevicesDeactivate_SendsDelete(t *testing.T) {
	var captured struct {
		method string
		path   string
	}
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		captured.method = r.Method
		captured.path = r.URL.Path
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"id":"dev-1","status":"inactive"}`))
	})

	result, err := client.Devices.Deactivate(context.Background(), "dev-1")
	if err != nil {
		t.Fatalf("Deactivate: %v", err)
	}
	if captured.method != "DELETE" {
		t.Errorf("method = %q, want DELETE", captured.method)
	}
	if captured.path != "/api/v1/devices/dev-1" {
		t.Errorf("path = %q", captured.path)
	}
	if result.ID != "dev-1" || result.Status != "inactive" {
		t.Errorf("got %+v", result)
	}
}

func TestDevicesDeactivate_404APIError(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"error":{"code":"not_found","message":"Device not found"}}`))
	})

	_, err := client.Devices.Deactivate(context.Background(), "missing")
	if err == nil {
		t.Fatal("expected APIError")
	}
	var apiErr *nitroping.APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if apiErr.StatusCode != 404 {
		t.Errorf("StatusCode = %d", apiErr.StatusCode)
	}
	if apiErr.Code != "not_found" {
		t.Errorf("Code = %q", apiErr.Code)
	}
}

func TestDevicesDeactivate_EmptyIDReturnsError(t *testing.T) {
	client, err := nitroping.NewClient(nitroping.ClientOptions{APIKey: "np_x"})
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	_, err = client.Devices.Deactivate(context.Background(), "")
	if err == nil {
		t.Fatal("expected error on empty device id")
	}
}

func TestDevicesList_GetsWithSnakeCaseQueryAndMapsRows(t *testing.T) {
	var captured struct {
		method string
		path   string
		query  url.Values
		auth   string
	}
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		captured.method = r.Method
		captured.path = r.URL.Path
		captured.query = r.URL.Query()
		captured.auth = r.Header.Get("Authorization")
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"data":[{"id":"dev-1","user_id":"alice","platform":"ios","status":"active","tags":["vip"],"timezone":"Europe/Istanbul","apns_environment":"production","last_seen_at":"2026-06-15T00:00:00Z","inserted_at":"2026-06-14T00:00:00Z"}],"total":1}`))
	})

	res, err := client.Devices.List(context.Background(), nitroping.ListDevicesQuery{
		UserID:   "alice",
		Platform: nitroping.PlatformIOS,
		PageSize: nitroping.Int(10),
	})
	if err != nil {
		t.Fatalf("List: %v", err)
	}

	if captured.method != "GET" {
		t.Errorf("method = %q, want GET", captured.method)
	}
	if captured.path != "/api/v1/devices" {
		t.Errorf("path = %q, want /api/v1/devices", captured.path)
	}
	if captured.auth != "ApiKey np_test_secret" {
		t.Errorf("Authorization = %q", captured.auth)
	}
	if captured.query.Get("user_id") != "alice" {
		t.Errorf("query user_id = %q, want alice", captured.query.Get("user_id"))
	}
	if captured.query.Get("platform") != "ios" {
		t.Errorf("query platform = %q, want ios", captured.query.Get("platform"))
	}
	if captured.query.Get("page_size") != "10" {
		t.Errorf("query page_size = %q, want 10", captured.query.Get("page_size"))
	}

	if res.Total != 1 {
		t.Errorf("Total = %d, want 1", res.Total)
	}
	if len(res.Data) != 1 {
		t.Fatalf("len(Data) = %d, want 1", len(res.Data))
	}
	d := res.Data[0]
	if d.ID != "dev-1" {
		t.Errorf("Data[0].ID = %q, want dev-1", d.ID)
	}
	if d.UserID == nil || *d.UserID != "alice" {
		t.Errorf("Data[0].UserID = %v, want alice", d.UserID)
	}
	if d.Platform != nitroping.PlatformIOS {
		t.Errorf("Data[0].Platform = %q, want ios", d.Platform)
	}
	if d.Status != "active" {
		t.Errorf("Data[0].Status = %q, want active", d.Status)
	}
	if len(d.Tags) != 1 || d.Tags[0] != "vip" {
		t.Errorf("Data[0].Tags = %v, want [vip]", d.Tags)
	}
	if d.Timezone == nil || *d.Timezone != "Europe/Istanbul" {
		t.Errorf("Data[0].Timezone = %v, want Europe/Istanbul", d.Timezone)
	}
	if d.APNsEnvironment == nil || *d.APNsEnvironment != "production" {
		t.Errorf("Data[0].APNsEnvironment = %v, want production", d.APNsEnvironment)
	}
	if d.LastSeenAt == nil || *d.LastSeenAt != "2026-06-15T00:00:00Z" {
		t.Errorf("Data[0].LastSeenAt = %v", d.LastSeenAt)
	}
	if d.InsertedAt != "2026-06-14T00:00:00Z" {
		t.Errorf("Data[0].InsertedAt = %q", d.InsertedAt)
	}
}

func TestDevicesList_NeverExposesToken(t *testing.T) {
	// The list endpoint never returns the push token; assert the
	// DeviceSummary type has no field that could surface one, by
	// round-tripping a body that (hypothetically) included a token and
	// confirming it is dropped rather than decoded into the struct.
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"data":[{"id":"dev-1","platform":"ios","status":"active","tags":[],"inserted_at":"2026-06-14T00:00:00Z","token":"SHOULD-NOT-APPEAR"}],"total":1}`))
	})

	res, err := client.Devices.List(context.Background(), nitroping.ListDevicesQuery{})
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if len(res.Data) != 1 {
		t.Fatalf("len(Data) = %d, want 1", len(res.Data))
	}

	// Re-marshal the decoded summary and confirm no token leaked through.
	raw, err := json.Marshal(res.Data[0])
	if err != nil {
		t.Fatalf("Marshal: %v", err)
	}
	var roundTrip map[string]any
	if err := json.Unmarshal(raw, &roundTrip); err != nil {
		t.Fatalf("Unmarshal: %v", err)
	}
	if _, ok := roundTrip["token"]; ok {
		t.Errorf("DeviceSummary unexpectedly carries a token field: %s", raw)
	}
}

func TestDevicesList_EmptyQueryHasNoQueryString(t *testing.T) {
	var rawQuery string
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		rawQuery = r.URL.RawQuery
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"data":[],"total":0}`))
	})

	res, err := client.Devices.List(context.Background(), nitroping.ListDevicesQuery{})
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if rawQuery != "" {
		t.Errorf("RawQuery = %q, want empty", rawQuery)
	}
	if res.Total != 0 || len(res.Data) != 0 {
		t.Errorf("got %+v, want empty listing", res)
	}
}

func TestDevicesDeactivateByToken_SendsTokenBody(t *testing.T) {
	var captured struct {
		method string
		path   string
		ctype  string
		body   map[string]any
	}
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		captured.method = r.Method
		captured.path = r.URL.Path
		captured.ctype = r.Header.Get("Content-Type")
		raw, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(raw, &captured.body)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(`{"id":"dev-9","status":"inactive"}`))
	})

	result, err := client.Devices.DeactivateByToken(context.Background(), "apns-token-xyz")
	if err != nil {
		t.Fatalf("DeactivateByToken: %v", err)
	}
	if captured.method != "DELETE" {
		t.Errorf("method = %q, want DELETE", captured.method)
	}
	if captured.path != "/api/v1/devices" {
		t.Errorf("path = %q, want /api/v1/devices", captured.path)
	}
	if captured.ctype != "application/json" {
		t.Errorf("Content-Type = %q", captured.ctype)
	}
	if captured.body["token"] != "apns-token-xyz" {
		t.Errorf("body.token = %v, want apns-token-xyz", captured.body["token"])
	}
	if result.ID != "dev-9" || result.Status != "inactive" {
		t.Errorf("got %+v, want {dev-9 inactive}", result)
	}
}

func TestDevicesDeactivateByToken_404APIError(t *testing.T) {
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"error":{"code":"not_found","message":"Device not found"}}`))
	})

	_, err := client.Devices.DeactivateByToken(context.Background(), "nope")
	if err == nil {
		t.Fatal("expected APIError")
	}
	var apiErr *nitroping.APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("expected *APIError, got %T: %v", err, err)
	}
	if apiErr.StatusCode != 404 {
		t.Errorf("StatusCode = %d", apiErr.StatusCode)
	}
	if apiErr.Code != "not_found" {
		t.Errorf("Code = %q", apiErr.Code)
	}
}

func TestDevicesDeactivateByToken_EmptyTokenReturnsError(t *testing.T) {
	client, err := nitroping.NewClient(nitroping.ClientOptions{APIKey: "np_x"})
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	_, err = client.Devices.DeactivateByToken(context.Background(), "")
	if err == nil {
		t.Fatal("expected error on empty token")
	}
}

func TestDevicesRegister_TimezoneForwarded(t *testing.T) {
	var captured map[string]any
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(raw, &captured)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"id":"dev-4","created":true}`))
	})

	_, err := client.Devices.Register(context.Background(), nitroping.DeviceRequest{
		Platform: nitroping.PlatformIOS,
		Token:    "apns-token",
		Timezone: nitroping.String("Europe/Istanbul"),
	})
	if err != nil {
		t.Fatalf("Register: %v", err)
	}
	if captured["timezone"] != "Europe/Istanbul" {
		t.Errorf("body.timezone = %v, want Europe/Istanbul", captured["timezone"])
	}
}

func TestDevicesRegister_TimezoneOmittedWhenUnset(t *testing.T) {
	var captured map[string]any
	client := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(raw, &captured)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"id":"dev-5","created":true}`))
	})

	_, err := client.Devices.Register(context.Background(), nitroping.DeviceRequest{
		Platform: nitroping.PlatformAndroid,
		Token:    "fcm-token",
	})
	if err != nil {
		t.Fatalf("Register: %v", err)
	}
	if _, ok := captured["timezone"]; ok {
		t.Errorf("body should not include timezone when unset")
	}
}

File: js/test/devices.test.ts
import { afterEach, describe, expect, it, vi } from "vitest"
import { Nitroping, NitropingError } from "../src/index"

function mockFetch(
  impl: (req: { url: string; init: RequestInit }) => Response | Promise<Response>,
) {
  return vi.spyOn(globalThis, "fetch").mockImplementation(async (input, init) => {
    const url =
      typeof input === "string"
        ? input
        : input instanceof URL
          ? input.toString()
          : (input as Request).url
    return await impl({ url, init: init ?? {} })
  })
}

describe("devices.register", () => {
  afterEach(() => vi.restoreAllMocks())

  it("posts to /api/v1/devices with snake_case body", async () => {
    const spy = mockFetch(
      () =>
        new Response(JSON.stringify({ id: "dev-1", created: true }), {
          status: 201,
          headers: { "Content-Type": "application/json" },
        }),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    const result = await np.devices.register({
      platform: "ios",
      token: "apns-token-abc",
      userId: "user-42",
      metadata: { source: "tests" },
    })

    expect(result).toEqual({ id: "dev-1", created: true })

    const init = spy.mock.calls[0]![1]!
    expect(init.method).toBe("POST")
    expect(spy.mock.calls[0]![0] as string).toBe("https://nitroping.dev/api/v1/devices")
    expect((init.headers as Record<string, string>)["Authorization"]).toBe("ApiKey np_x")

    const body = JSON.parse(init.body as string)
    expect(body).toEqual({
      platform: "ios",
      token: "apns-token-abc",
      user_id: "user-42",
      metadata: { source: "tests" },
    })
  })

  it("supports the web platform with p256dh/auth keys", async () => {
    const spy = mockFetch(
      () =>
        new Response(JSON.stringify({ id: "dev-2", created: false }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    const result = await np.devices.register({
      platform: "web",
      token: "https://fcm.googleapis.com/abc",
      webPushP256dh: "BPS_p256dh_value",
      webPushAuth: "auth_secret_value",
    })

    expect(result.created).toBe(false)
    const body = JSON.parse(spy.mock.calls[0]![1]!.body as string)
    expect(body.web_push_p256dh).toBe("BPS_p256dh_value")
    expect(body.web_push_auth).toBe("auth_secret_value")
  })

  it("forwards the iOS environment field on the wire", async () => {
    const spy = mockFetch(
      () =>
        new Response(JSON.stringify({ id: "dev-3", created: true }), {
          status: 201,
          headers: { "Content-Type": "application/json" },
        }),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    await np.devices.register({
      platform: "ios",
      token: "ios-token",
      environment: "sandbox",
    })

    const body = JSON.parse(spy.mock.calls[0]![1]!.body as string)
    expect(body.environment).toBe("sandbox")
  })

  it("forwards the timezone field on the wire", async () => {
    const spy = mockFetch(
      () =>
        new Response(JSON.stringify({ id: "dev-tz", created: true }), {
          status: 201,
          headers: { "Content-Type": "application/json" },
        }),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    await np.devices.register({
      platform: "ios",
      token: "ios-token",
      timezone: "Europe/Istanbul",
    })

    const body = JSON.parse(spy.mock.calls[0]![1]!.body as string)
    expect(body.timezone).toBe("Europe/Istanbul")
  })

  it("omits environment when not provided", async () => {
    const spy = mockFetch(
      () =>
        new Response(JSON.stringify({ id: "dev-4", created: true }), {
          status: 201,
          headers: { "Content-Type": "application/json" },
        }),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    await np.devices.register({ platform: "ios", token: "ios-token" })

    const body = JSON.parse(spy.mock.calls[0]![1]!.body as string)
    expect("environment" in body).toBe(false)
  })
})

describe("devices.deactivate", () => {
  afterEach(() => vi.restoreAllMocks())

  it("sends DELETE to /api/v1/devices/:id", async () => {
    const spy = mockFetch(
      () =>
        new Response(JSON.stringify({ id: "dev-1", status: "inactive" }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    const result = await np.devices.deactivate("dev-1")

    expect(result).toEqual({ id: "dev-1", status: "inactive" })
    expect(spy.mock.calls[0]![0] as string).toBe("https://nitroping.dev/api/v1/devices/dev-1")
    expect(spy.mock.calls[0]![1]!.method).toBe("DELETE")
  })

  it("throws NitropingError with not_found on 404", async () => {
    mockFetch(
      () =>
        new Response(
          JSON.stringify({ error: { code: "not_found", message: "Device not found" } }),
          { status: 404, headers: { "Content-Type": "application/json" } },
        ),
    )

    const np = new Nitroping({ apiKey: "np_x" })

    try {
      await np.devices.deactivate("missing")
      expect.fail("expected error")
    } catch (err) {
      const e = err as NitropingError
      expect(e).toBeInstanceOf(NitropingError)
      expect(e.code).toBe("not_found")
      expect(e.status).toBe(404)
    }
  })
})

describe("devices.list", () => {
  afterEach(() => vi.restoreAllMocks())

  it("GETs /api/v1/devices with snake_case query and camelCases the rows", async () => {
    const spy = mockFetch(
      () =>
        new Response(
          JSON.stringify({
            data: [
              {
                id: "dev-1",
                user_id: "alice",
                platform: "ios",
                status: "active",
                tags: ["vip"],
                timezone: "Europe/Istanbul",
                apns_environment: "production",
                last_seen_at: "2026-06-15T00:00:00Z",
                inserted_at: "2026-06-14T00:00:00Z",
              },
            ],
            total: 1,
          }),
          { status: 200, headers: { "Content-Type": "application/json" } },
        ),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    const res = await np.devices.list({ userId: "alice", platform: "ios", pageSize: 10 })

    expect(res.total).toBe(1)
    expect(res.data[0]).toEqual({
      id: "dev-1",
      userId: "alice",
      platform: "ios",
      status: "active",
      tags: ["vip"],
      timezone: "Europe/Istanbul",
      apnsEnvironment: "production",
      lastSeenAt: "2026-06-15T00:00:00Z",
      insertedAt: "2026-06-14T00:00:00Z",
    })

    const url = spy.mock.calls[0]![0] as string
    expect(url).toContain("/api/v1/devices?")
    expect(url).toContain("user_id=alice")
    expect(url).toContain("platform=ios")
    expect(url).toContain("page_size=10")
    expect(spy.mock.calls[0]![1]!.method).toBe("GET")
  })

  it("returns an empty listing as { data: [], total: 0 }", async () => {
    mockFetch(
      () =>
        new Response(JSON.stringify({ data: [], total: 0 }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    const res = await np.devices.list()
    expect(res).toEqual({ data: [], total: 0 })
  })
})

describe("devices.deactivateByToken", () => {
  afterEach(() => vi.restoreAllMocks())

  it("DELETEs /api/v1/devices with a { token } body", async () => {
    const spy = mockFetch(
      () =>
        new Response(JSON.stringify({ id: "dev-9", status: "inactive" }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    const res = await np.devices.deactivateByToken("apns-token-xyz")

    expect(res).toEqual({ id: "dev-9", status: "inactive" })
    expect(spy.mock.calls[0]![0] as string).toBe("https://nitroping.dev/api/v1/devices")
    const init = spy.mock.calls[0]![1]!
    expect(init.method).toBe("DELETE")
    expect(JSON.parse(init.body as string)).toEqual({ token: "apns-token-xyz" })
  })

  it("throws not_found when no device matches the token", async () => {
    mockFetch(
      () =>
        new Response(
          JSON.stringify({ error: { code: "not_found", message: "Device not found" } }),
          { status: 404, headers: { "Content-Type": "application/json" } },
        ),
    )

    const np = new Nitroping({ apiKey: "np_x" })
    await expect(np.devices.deactivateByToken("nope")).rejects.toMatchObject({ code: "not_found" })
  })
})
